File: haskell/lib/FpInScala/Part1/Chapter3/FunctionalDataStructure.hs
module FpInScala.Part1.Chapter3.FunctionalDataStructure
  ( List(Nil, Cons)
  , sumList
  , list
  , tail'
  , setHead
  , drop'
  , dropWhile'
  , init'
  , foldRight
  , length'
  , foldLeft
  , sumLeft
  , productLeft
  , lengthLeft
  , reverse'
  , foldLeftFromRight
  , foldRightFromLeft
  , appendRight
  , concat'
  , incrementEach
  , doubleToString
  , map'
  , filter'
  , flatMap
  , filterFromFlatMap
  , addPairwise
  ) where

data List a = Nil | Cons a (List a)

instance Show a => Show (List a) where
  show Nil = "Nil"
  show (Cons x xs) = "Cons " ++ show x ++ " (" ++ show xs ++ ")"

instance Eq a => Eq (List a) where
  Nil == Nil = True
  (Cons x xs) == (Cons y ys) = x == y && xs == ys
  _ == _ = False

sumList :: List Int -> Int
sumList Nil = 0
sumList (Cons x xs) = x + sumList xs

productList :: List Double -> Double
productList Nil = 1.0
productList (Cons 0.0 _) = 0.0
productList (Cons x xs) = x * productList xs

list :: [a] -> List a
list [] = Nil
list (x:xs) = Cons x (list xs)

tail' :: List a -> List a
tail' Nil = error "Nil"
tail' (Cons _ xs) = xs

setHead :: List a -> a -> List a
setHead Nil _ = error "Nil"
setHead (Cons _ xs) x = Cons x xs

drop' :: List a -> Int -> List a
drop' Nil _ = Nil
drop' list'@(Cons _ xs) n
  | n <= 0 = list'
  | otherwise = drop' xs (n - 1)

dropWhile' :: List a -> (a -> Bool) -> List a
dropWhile' Nil _ = Nil
dropWhile' list'@(Cons x xs) f
  | f x = dropWhile' xs f
  | otherwise = list'

append :: List a -> List a -> List a
append Nil a2 = a2
append (Cons h t) a2 = Cons h (append t a2)

init' :: List a -> List a
init' Nil = error "Nil"
init' (Cons x Nil) = Nil
init' (Cons x xs) = Cons x (init' xs)

foldRight :: List a -> b -> (a -> b -> b) -> b
foldRight Nil b _ = b
foldRight (Cons x xs) b f = f x (foldRight xs b f)

length' :: List a -> Int
length' xs = foldRight xs 0 (\a b -> b + 1)

foldLeft :: List a -> b -> (b -> a -> b) -> b
foldLeft Nil b _ = b
foldLeft (Cons x xs) b f = foldLeft xs (f b x) f

sumLeft :: List Int -> Int
sumLeft xs = foldLeft xs 0 (\a x -> a + x)

productLeft :: List Double -> Double
productLeft xs = foldLeft xs 1.0 (\a x -> a * x)

lengthLeft :: List a -> Int
lengthLeft xs = foldLeft xs 0 (\a _ -> a + 1)

reverse' :: List a -> List a
reverse' xs = foldLeft xs Nil (\a head -> Cons head a)

foldLeftFromRight :: List a -> b -> (b -> a -> b) -> b
foldLeftFromRight xs b f = foldRight xs b (\b h -> f h b)

foldRightFromLeft :: List a -> b -> (a -> b -> b) -> b
foldRightFromLeft xs b f = foldLeft xs b (\h b -> f b h)

appendRight :: List a ->  List a -> List a
appendRight a1 a2 = foldRight a1 a2 (\h b -> Cons h b)

concat' :: List (List a) -> List a
concat' xss = foldRight xss Nil append

incrementEach :: List Int -> List Int
incrementEach Nil = Nil
incrementEach xs = foldRight xs Nil (\h acc -> Cons (h + 1) acc)

doubleToString :: List Double -> List String
doubleToString xs = foldRight xs Nil (\h acc -> Cons (show h) acc)

map' :: List a -> (a -> b) ->List b
map' xs f = foldRight xs Nil (\h acc -> Cons (f h) acc)

filter' :: List a -> (a -> Bool) -> List a
filter' xs f = foldRight xs Nil (\h acc -> if f h then Cons h acc else acc)

flatMap :: List a -> (a -> List b) -> List b
flatMap xs f = concat' (map' xs f)

filterFromFlatMap :: List a -> (a -> Bool) -> List a
filterFromFlatMap xs f = flatMap xs  (\x -> if f x then list [x] else list [] )

addPairwise :: List Int -> List Int -> List Int
addPairwise Nil _ = Nil
addPairwise _ Nil = Nil
addPairwise (Cons a1 as1) (Cons a2 as2) = Cons (a1 + a2) (addPairwise as1 as2)

File: scala/src/test/scala/part1/chapter3/FunctionalDataStructure.scala
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.Assertions._

import part1.chapter3.*
import part1.chapter3.List.apply

class FunctionalDataStructureSuite extends munit.FunSuite {
  test("Exercise 3.1") {
    val result = List(1,2,3,4,5) match
        case List.Cons(x, List.Cons(2, List.Cons(4, _))) => x
        case List.Nil => 42
        case List.Cons(x, List.Cons(y, List.Cons(3, List.Cons(4, _)))) => x + y
        case List.Cons(h, t) => h + List.sum(t)
        // case _ => 101
    assertEquals(result, 3)
  }

  test("Exercise 3.2") {
    assertThrows[Exception] {
      List.tail(List())
    }
    assertEquals(List.tail(List(1,2,3)), List(2,3))
  }

  test("Exercise 3.3") {
    assertThrows[Exception] {
      List.setHead(List[Int](), 42)
    }
    assertEquals(List.setHead(List(1,2,3), 42), List(42, 2,3))
  }

  test("Exercise 3.4") {
    assertEquals(List.drop(List(), 42), List())
    assertEquals(List.drop(List(1,2,3,4,5), 3), List(4,5))
  }

  test("Exercise 3.5") {
    assertEquals(List.dropWhile(List(), (n: Int) => n < 3), List())
    assertEquals(List.dropWhile(List(1,2,3,4,5), (n: Int) => n < 3), List(3, 4, 5))
  }

  test("Exercise 3.6") {
    assertThrows[Exception] {
      List.init(List())
    }
    assertEquals(List.init(List(1,2,3,4,5)), List(1, 2, 3, 4))
  }

  test("Exercise 3.8") {
    assertEquals(List.foldRight(List(1, 2, 3), List.Nil, (l, r: List[Int] ) => List.Cons(l, r)), List(1, 2, 3))
  }

  test("Exercise 3.9") {
    assertEquals(List.length(List.Nil), 0)
    assertEquals(List.length(List(1, 2, 3)), 3)
  }

  test("Exercise 3.10") {
    assertEquals(List.foldLeft(List(1, 2, 3), 0, _ + _), 6)
  }

  test("Exercise 3.11") {
    assertEquals(List.sumLeft(List(1, 2, 3)), 6)
    assertEquals(List.productLeft(List(1.0, 2.0, 3.0)), 6.0)
    assertEquals(List.lengthLeft(List(1, 2, 3)), 3)
  }

  test("Exercise 3.12") {
    assertEquals(List.reverse(List(1, 2, 3)), List(3, 2, 1))
  }

  test("Exercise 3.13") {
    assertEquals(List.foldLeftFromRight(List(1, 2, 3), 0, _ + _), 6)
    assertEquals(List.foldRightFromLeft(List(1, 2, 3), 0, _ + _), 6)
  }

  test("Exercise 3.14") {
    assertEquals(List.appendRight(List(1, 2, 3), List(4, 5, 6)), List(1, 2, 3, 4, 5, 6))
  }

  test("Exercise 3.15") {
    assertEquals(List.concat(List(List(1, 2, 3), List(4, 5, 6))), List(1, 2, 3, 4, 5, 6))
  }

  test("Exercise 3.16") {
    assertEquals(List.incrementEach(List(1, 2, 3)), List(2, 3, 4))
  }

  test("Exercise 3.17") {
    assertEquals(List.doubleToString(List(1.0, 2.0, 3.0, 4.0, 5.0)), List("1.0", "2.0", "3.0", "4.0", "5.0"))
  }

  test("Exercise 3.18") {
    assertEquals(List.map(List(1, 2, 3, 4, 5), _ * 2), List(2, 4, 6, 8, 10))
  }

  test("Exercise 3.19") {
    assertEquals(List.filter(List(1, 2, 3, 4, 5), (x) => x % 2 != 0), List(1, 3, 5))
  }

  test("Exercise 3.20") {
    assertEquals(List.flatMap(List(1, 2, 3), (x) => List(x, x)), List(1, 1, 2, 2, 3, 3))
  }

  test("Exercise 3.21") {
    assertEquals(List.filterFromFlatMap(List(1, 2, 3), (x) => x % 2 != 0), List(1, 3))
  }

  test("Exercise 3.22") {
    assertEquals(List.addPairwise(List(1, 2, 3), List(4, 5, 6)), List(5, 7, 9))
  }

  test("Exercise 3.23") {
    assertEquals(List.zipWith(List(1, 2, 3), List(4, 5, 6), _ + _), List(5, 7, 9))
  }

  test("LISTS IN THE STANDARD LIBRARY") {
    assertEquals((List.take(List(1, 2, 3, 4, 5), 3)), List(1, 2, 3))
    assertEquals((List.takeWhile(List(1, 2, 3, 4, 5), _ <= 3)), List(1, 2, 3))
    assertEquals(List.forall(List(1, 2, 3, 4, 5), _ <= 3), false)
    assertEquals(List.forall(List(1, 2, 3, 4, 5), _ <= 5), true)
    assertEquals(List.exists(List(1, 2, 3, 4, 5), _ == 42), false)
    assertEquals(List.exists(List(1, 2, 3, 4, 5), _ == 3), true)
    assertEquals(List.scanLeft(List("a", "b", "c", "d", "e"), "", _ + _), List("", "a", "ab", "abc", "abcd", "abcde"))
    assertEquals(List.scanRight(List("a", "b", "c", "d", "e"), "", _ + _), List("abcde", "bcde", "cde", "de", "e", ""))
  }

  test("Exercise 3.24") {
    assertEquals(List.hasSubsequence(List(1, 2, 3, 4, 5), List(5, 4, 3)), false)
    assertEquals(List.hasSubsequence(List(1, 2, 3, 4, 5), List(2, 3, 4)), true)
  }

}
